flowinfra: replace ConcurrentExecution with ConcurrentTxnUse

asubiotto opened the pull request:
Previously, the DistSQLPlanner would reject any local flows with any
concurrency due to possible concurrent use of a RootTxn. This policy was overly
prohibitive as it would reject stats flows that contain processors that cannot
be fused even though there would be no concurrent txn usage.

This commit makes this policy a bit more lenient, introducing a way for
processors to indicate that they and their input will not be using a txn. This
is exposed through the new ConcurrentTxnUse method in the flow interface, which
returns whether more than one processor will be using a txn.

This commit also removes ConcurrentExecution since all usages have been
replaced with ConcurrentTxnUse.

Release note: None (internal refactor to fix an error message received only in
a multitenant environment)

Fixes #50049

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52621)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Are there any other processors that cannot be fused but also don't use a txn?

> Reviewed 13 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52621#-:-METpb1qCiDWFc37vsXd:b-v3u429)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @asubiotto)

___
*[pkg/sql/execinfra/processorsbase.go, line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52621#-METopOI86TGU0CaSYe0:-METopOI86TGU0CaSYe1:b-lrmeng) ([raw file](https://github.com/cockroachdb/cockroach/blob/284cbb5addedbff1a6b0a1db2384d206e1bc4596/pkg/sql/execinfra/processorsbase.go#L45)):*
> ```Go
> // local flow from running in parallel if this is unknown since concurrent use
> // of the RootTxn is forbidden (in a distributed flow these are leaf txns, so
> // it doesn't matter.)
> ```

nit: period before parenthesis.


<!-- Sent from Reviewable.io -->

Commit:
flowinfra: replace ConcurrentExecution with ConcurrentTxnUse

Previously, the DistSQLPlanner would reject any local flows with any
concurrency due to possible concurrent use of a RootTxn. This policy was overly
prohibitive as it would reject stats flows that contain processors that cannot
be fused even though there would be no concurrent txn usage.

This commit makes this policy a bit more lenient, introducing a way for
processors to indicate that they and their input will not be using a txn. This
is exposed through the new ConcurrentTxnUse method in the flow interface, which
returns whether more than one processor will be using a txn.

This commit also removes ConcurrentExecution since all usages have been
replaced with ConcurrentTxnUse.

Release note: None (internal refactor to fix an error message received only in
a multitenant environment)

asubiotto reviewed (commented):
Probably, but it's not the goal of this PR to implement `DoesNotUseTxn` for all processors, only for the stats ones to fix the immediate problem of not being able to run `CREATE STATS` in a multitenant environment. If we encounter a problem related to other processors, the fix should be as easy as implementing `DoesNotUseTxn`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52621#-:-MEWlKnV-H4PFGZz94Xr:b-g7ovoc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52621#-:-MEYF37GEOnkdOd8JKcD:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> that cannot be fused even though there would be no concurrent txn usage

Is there a good reason why whatever prompted this cannot be fused?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52621#-:-MEYGq3Z53nPb0802oNz:bekz1w)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)



<!-- Sent from Reviewable.io -->

asubiotto commented:
It's another solution I mentioned in (https://github.com/cockroachdb/cockroach/issues/50049#issuecomment-671858358), but it's code I'm not familiar with (stats collection processors) and will probably be non-trivial based on my experience refactoring so want to leave it on the optimizer team's plate. Created an issue: https://github.com/cockroachdb/cockroach/issues/52700

andreimatei reviewed (commented):
Let's see what @rytaft says. If we can just fuse whatever, I'd rather do that than introduce this interface since I don't trust an interface like this to stay accurate.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52621#-:-MEYbnL0AxsSMatyxXub:b-oltmh8)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
@rytaft is OOO this week. I don't see any reason we can't fuse them, but we are all pretty swamped right now.

asubiotto commented:
To be clear: the inability to fuse in the commit message is in reference to the current state of the DistSQLPlanner not being able to do so because the processors do not implement `RowSource`, not that these processors cannot implement `RowSource`, which would make them fusable.

I'm going to go ahead and merge this to unblock stats creation in a multitenant environment. This is an opt-in interface, and I think it's very unlikely for any processors that need to implement this interface to change their behavior to suddenly use a `txn` since it's such an intrinsic part of the processor. Also, the number of processors that don't implement `RowSource` is probably pretty low. Once #52700 is dealt with, I'm happy to remove this interface.

bors r=yuzefovich
